lib/puppet/pops/types/p_binary_type.rb

```diff
@@ -0,0 +1,194 @@
+    # the string is broken up with new lines.
+    def relaxed_to_s
+      Base64.encode64(@binary_buffer)
+    end
+
+    # Returns the binary content as a url safe base64 string (where + and / are replaced by - and _)
+    #
+    def urlsafe_to_s
+      Base64.urlsafe_encode64(@binary_buffer)
+    end
+
+    def eql?(o)
+      self.class == o.class && @binary_buffer == o.binary_buffer
+    end
+
+    def ==(o)
+      self.eql?(o)
+    end
+
+    def length()
+      @binary_buffer.length
+    end
+  end
+
+  def self.register_ptype(loader, ir)
+    create_ptype(loader, ir, 'AnyType')
+  end
+
+  def initialize()
+  end
+
+  # Only instances of Binary are instances of the PBinaryType
+  #
+  def instance?(o, guard = nil)
+    o.is_a?(Binary)
+  end
+
+  def eql?(o)
+    self.class == o.class
+  end
+
+  def hash?
+    super
+  end
+
+  # @api private
+  def self.new_function(_, loader)
+    @new_function ||= Puppet::Functions.create_loaded_function(:new_Binary, loader) do
+      local_types do
+        type 'ByteInteger = Integer[0,255]'
+        type 'Base64Format = Enum["%b", "%u", "%B", "%s"]'
+        type 'StringHash = Struct[{value => String, "format" => Optional[Base64Format]}]'
+        type 'ArrayHash = Struct[{value => Array[ByteInteger]}]'
+        type 'BinaryArgsHash = Variant[StringHash, ArrayHash]'
+      end
+
+      # Creates a binary from a base64 encoded string in one of the formats %b (
+      dispatch :from_string do
+        param 'String', :str
+        optional_param 'Base64Format', :format
+      end
+
+      dispatch :from_array do
+        param 'Array[ByteInteger]', :byte_array
+      end
+
+      # Same as from_string, or from_array, but value and (for string) optional format are given in the form
+      # of a hash.
+      #
+      dispatch :from_hash do
+        param 'BinaryArgsHash', :hash_args
+      end
+
+      def from_string(str, format = nil)
+        format ||= '%b'
+        case format
+        when "%b"
+          Binary.new(Base64.decode64(str))
+
+        when "%u"
+          Binary.new(Base64.urlsafe_decode64(str))
+
+        when "%B"
+          Binary.new(Base64.strict_decode64(str))
+
+        when "%s"
+          Binary.new(str)
+
+        else
+          raise ArgumentError, "Unsupported Base64 format '#{format}'"
+        end
+      end
+
+      def from_array(array)
+        # The array is already known to have bytes in the range 0-255, or it is in error
+        # Without this pack C would produce weird results
+        Binary.from_binary_string(array.pack("C*"))
+      end
+
+      def from_hash(hash)
+        case hash['value']
+        when Array
+          from_array(hash['value'])
+        when String
+          from_string(hash['value'], hash['format'])
+        end
+      end
+    end
+  end
+
+  DEFAULT = PBinaryType.new
+
+  protected
+
+  def _assignable?(o, guard)
+    o.class == self.class
+  end
+
+end
```

Comment:
Abandoned. Bad idea. A Binary is simply a binary buffer and any encoding is about how it is presented in some medium.